### PR TITLE
Allow using path expressions in paginator outputs

### PR DESCRIPTION
**Pull request by JordonPhillips**
*Description of changes:*

This allows the use of path expressions in paginator outputs. This enables
customers to have result wrapper shapes, which is a common pattern in XML
protocols.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/source/spec/core.rst [hunk 1]
```diff
@@ -4045,6 +4045,112 @@ settings from a service.
             }
         }
 
+The values for ``outputToken`` and ``items`` are paths. Paths are a series of
```
Comment (mtdowling) [inline docs/source/spec/core.rst]: Why do we not support paths for inputToken and pageSize?

File: docs/source/spec/core.rst [hunk 2]
```diff
@@ -4045,6 +4045,112 @@ settings from a service.
...
+member name in a structure. The first member name MUST correspond to a member
+of the output structure and each subsequent member name MUST correspond to a
+member in the previously referenced structure. Paths MUST adhere to the
+following ABNF.
+
+.. productionlist:: smithy
+    path    :`identifier` *("." `identifier`)
+
+The following example defines a paginated operation which uses a result
+wrapper where the output token and items are indicated with paths.
```
Comment (mtdowling) [inline docs/source/spec/core.rst]: Minor: would this read better? "indicated with" -> "referenced by"

File: docs/source/spec/core.rst (hunk 1 above)
Comment (mtdowling) [inline docs/source/spec/core.rst]: Minor: can you put italics on "Path" to show it's defining a term?

File: docs/source/spec/core.rst [hunk 3]
```diff
@@ -4045,6 +4045,112 @@ settings from a service.
...
+
+.. tabs::
+
+    .. code-tab:: smithy
+
+        namespace smithy.example
+
+        @collection @readonly
```
Comment (mtdowling) [inline docs/source/spec/core.rst]: The `@collection` trait no longer exists

File: docs/source/spec/core.rst [hunk 4]
```diff
@@ -4045,6 +4045,112 @@ settings from a service.
...
+        {
+            "smithy": "0.4.0",
+            "smithy.example": {
+                "shapes": {
+                    "GetFoos": {
+                        "type": "operation",
+                        "input" :"GetFoosInput",
+                        "output": "GetFoosOutput",
+                        "readonly": true,
+                        "collection": true,
```
Comment (mtdowling) [inline docs/source/spec/core.rst]: Remove

File: smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/PaginatedTraitValidator.java [hunk 5]
```diff
@@ -120,20 +122,26 @@
...
-        if (memberName == null) {
+        if (memberPath == null) {
             return service != null && validator.isRequiredToBePresent()
                    ? Collections.singletonList(error(operation, trait, String.format(
                     "%spaginated trait `%s` is not configured", prefix, validator.propertyName())))
                    : Collections.emptyList();
         }
 
-        MemberShape member = validator.getMember(opIndex, operation, trait).orElse(null);
+        if (!validator.pathsAllowed() && memberPath.split("\\.").length != 1) {
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/PaginatedTraitValidator.java]: Minor: Storing the regex in a constant means we don't need to re-compile it each time. Could you just use `memberPath.contains(".")`?

File: docs/source/spec/core.rst [hunk 6]
```diff
@@ -4045,6 +4045,112 @@ settings from a service.
...
 
+The values for ``outputToken`` and ``items`` are paths. Paths are a series of
+identifiers separated by dots (``.``) where each identifier represents a
+member name in a structure. The first member name MUST correspond to a member
+of the output structure and each subsequent member name MUST correspond to a
+member in the previously referenced structure. Paths MUST adhere to the
+following ABNF.
+
+.. productionlist:: smithy
+    path    :`identifier` *("." `identifier`)
```
Comment (mtdowling) [inline docs/source/spec/core.rst]: Why limit this to two levels deep?

File: smithy-model/src/test/java/software/amazon/smithy/model/traits/PaginatedTraitTest.java [hunk 7]
```diff
@@ -97,4 +97,35 @@ public void mergesWithOtherTrait() {
...
+
+    @Test
+    public void allowsNestedOutputToken() {
+        TraitFactory provider = TraitFactory.createServiceFactory();
+        Node node = Node.objectNode()
+                .withMember("inputToken", Node.from("inputToken"))
+                .withMember("outputToken", Node.from("result.outputToken"));
+
+        Optional<Trait> trait = provider.createTrait(
+                ShapeId.from("smithy.api#paginated"), ShapeId.from("ns.qux#foo"), node);
```
Comment (mtdowling) [inline smithy-model/src/test/java/software/amazon/smithy/model/traits/PaginatedTraitTest.java]: Minor: we typically use a newline to separate test setup from test assertions

File: docs/source/spec/core.rst (hunk 6 above)
Comment (JordonPhillips) [reply docs/source/spec/core.rst]: It isn't limited to two levels deep - you'll get a warning if it's more than that but your build will still pass. The warning exists because every level of nesting makes the api progressively more cumbersome to use. In an ideal world we would only have the result list and the output token in the response - anything else makes it hard to provide useful abstractions like iterators. A single level of nesting can be justified if it's a result wrapper shape, which is a fairly common pattern.

File: docs/source/spec/core.rst (hunk 1 above)
Comment (JordonPhillips) [reply docs/source/spec/core.rst]: We can always add support for this later, but it wasn't added because it makes the operations significantly more cumbersome to use. For reference, 0 of the several hundred existing AWS paginators use paths for input.

Comment (kstich) [reply docs/source/spec/core.rst]: Should use the `:dfn:` directive.

File: smithy-model/src/test/resources/software/amazon/smithy/model/errorfiles/validators/paginated-trait-test.errors [hunk 8]
```diff
@@ -18,3 +18,7 @@
 [ERROR] ns.foo#UnresolvedOutput$nextToken: member shape targets an unresolved shape `ns.foo#Missing` | Target
 [ERROR] ns.foo#Invalid9: When bound within the `ns.foo#Service` service, paginated trait `inputToken` is not configured | PaginatedTrait
 [ERROR] ns.foo#Invalid9: When bound within the `ns.foo#Service` service, paginated trait `outputToken` is not configured | PaginatedTrait
+[WARNING] ns.foo#DeeplyNestedOutputOperation: paginated trait `items` should not include a path with more than two parts | PaginatedTrait
```
Comment (mtdowling) [inline smithy-model/src/test/resources/software/amazon/smithy/model/errorfiles/validators/paginated-trait-test.errors]: Since this is the reference implementation of all things Smithy, this guidance here comes off almost as giving permission to implementers that more than 2 levels deep isn't supported. Can you try to rephrase this to be more about avoiding more than 2 levels to improve API quality?

File: docs/source/spec/core.rst (hunk 6 above)
Comment (mtdowling) [reply docs/source/spec/core.rst]: Ok, that sounds reasonable.

File: docs/source/spec/core.rst (hunk 1 above)
Comment (mtdowling) [reply docs/source/spec/core.rst]: 👌

File: smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/PaginatedTraitValidator.java [hunk 9]
```diff
@@ -59,6 +62,7 @@
     private static final Set<ShapeType> ITEM_SHAPES = SetUtils.of(ShapeType.LIST, ShapeType.MAP);
     private static final Set<ShapeType> PAGE_SHAPES = SetUtils.of(ShapeType.INTEGER);
     private static final Set<ShapeType> STRING_SET = SetUtils.of(ShapeType.STRING);
+    private static final java.util.regex.Pattern PATH_PATTERN = Pattern.compile("\\.");
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/PaginatedTraitValidator.java]: Whoops - can you remove this fully-qualified reference?

File: smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/PaginatedTraitValidator.java [hunk 10]
```diff
@@ -164,9 +182,53 @@
...
+        @Override
+        boolean pathsAllowed() {
+            return true;
+        }
+
+        Optional<MemberShape> getMember(
+                ShapeIndex index, OperationIndex opIndex, OperationShape operation, PaginatedTrait trait
+        ) {
+            // Split up the path expression into a list of member names
+            Optional<List<String>> memberNames = getMemberPath(opIndex, operation, trait)
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/PaginatedTraitValidator.java]: Minor: this could be simplified to:

```java
Optional<List<String>> memberNames = getMemberPath(opIndex, operation, trait)
        .map(value -> Arrays.asList(value.split("\\.")))
        .orElse(Collections.emptyList());
```
